addons/mail/tests/test_ir_mail_server.py

```diff
@@ -441,3 +444,253 @@ def test_personal_mail_server(self):
         self.assertFalse(self.emails)
+
+    @mute_logger('odoo.models.unlink')
+    def test_personal_mail_server_limit(self):
+        # Test the limit per personal mail servers
+        self.env['ir.config_parameter'].set_param('mail.personal_server_limit_per_minute', '20')
+        user_1, user_2 = self.env['res.users'].create([{
+            'name': f'User {i}',
+            'login': f'user_{i}_test',
+            'email': f'user_{i}_test@test.com',
+            'group_ids': [(6, 0, [self.env.ref('base.group_user').id])],
+        } for i in range(2)])
+
+        mail_server_1, __ = self.env["ir.mail_server"].create([{
+            'name': 'test',
+            'owner_user_id': user.id,
+            'from_filter': user.email,
+            'smtp_user': user.email,
+            'smtp_host': 'test@example.com',
+        } for user in (user_1, user_2)])
+
+        with self._freeze_time("2025-01-01 20:02:23"):
+            mails_user_1 = self.env["mail.mail"].with_user(user_1).sudo().create([
+                {'state': 'outgoing', 'email_to': 'target@test.com', 'email_from': user_1.email}
+                for i in range(82)
+            ])
+            mails_user_2 = self.env["mail.mail"].with_user(user_2).sudo().create([
+                {'state': 'outgoing', 'email_to': 'target@test.com', 'email_from': user_2.email}
+                for i in range(77)
+            ])
+            mails_other = self.env["mail.mail"].create([
+                {'state': 'outgoing', 'email_to': 'target@test.com', 'email_from': user_1.email}
+                for i in range(100)
+            ])
+
+        mails = mails_other + mails_user_1 + mails_user_2
+
+        self.assertEqual(mail_server_1.owner_limit_count, 0)
+        self.assertFalse(mail_server_1.owner_limit_time)
+
+        with self.mock_smtplib_connection(), self._freeze_time("2025-01-01 20:05:23"):
+            mails.send()
+
+        self.assertEqual(mail_server_1.owner_limit_count, 20)
+        self.assertEqual(mail_server_1.owner_limit_time, datetime(2025, 1, 1, 20, 5, 23))
+
+        self.assertEqual(self.connect_mocked.call_count, 3, "Called once for each mail server")
+
+        # Should have sent 20 emails, and delayed the other
+        self.assertEqual(self.env["mail.mail"].search_count(
+            [('id', 'in', mails_user_1.ids), ('state', '=', 'sent')]), 20)
+        delayed = self.env["mail.mail"].search(
+            [('id', 'in', mails_user_1.ids), ('state', '!=', 'sent')])
+        self.assertEqual(len(delayed), 82 - 20)
+        scheduled_dates = delayed.mapped('scheduled_date')
+        self.assertTrue(all(scheduled_dates))
+        scheduled_dates = sorted(scheduled_dates)
+
+        # Check the emails of the first user
+        # First 20 delayed will be sent in 1 minutes
+        # Next 20 delayed will be sent in 2 minutes
+        # ...
+        self.assertEqual(set(scheduled_dates[:20]), {datetime(2025, 1, 1, 20, 6, 23)})
```

Comment:
Same here, should probably check for the 3 servers to be sure we are server-batch-enabled, and that we did not just shorcut after finishing the first batch-server